Support round caps for the fast arc stroke generator

b-luk opened the pull request:
<!--
Thanks for filing a pull request!
Reviewers are typically assigned within a week of filing a request.
To learn more about code review, see our documentation on Tree Hygiene: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
-->

Support round caps for the fast arc stroke generator.

Factors out `Tessellator::GenerateRoundCapLine`'s cap generation logic into a new `Tessellator::GenerateRoundCap` method to be shared by `Tessellator::GenerateRoundCapLine` and `Tessellator::GenerateStrokedArc`.

Fixes #169400

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

**Note**: The Flutter team is currently trialing the use of [Gemini Code Assist for GitHub](https://developers.google.com/gemini-code-assist/docs/review-github-code). Comments from the `gemini-code-assist` bot should not be taken as authoritative feedback from the Flutter team. If you find its comments useful you can update your code accordingly, but if you are unsure or disagree with the feedback, please feel free to wait for a Flutter team member's review for guidance on which automated comments should be addressed.

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

b-luk committed:
Support round caps for the fast arc stroke generator

flar commented on engine/src/flutter/impeller/tessellator/tessellator.cc:
```diff
@@ -576,9 +584,17 @@ ArcVertexGenerator Tessellator::StrokedArc(const Matrix& view_transform,
+  std::unique_ptr<Trigs> round_cap_trigs;
+  if (cap == Cap::kRound) {
+    size_t round_cap_divisions = ComputeQuadrantDivisions(
+        view_transform.GetMaxBasisLengthXY() * half_width);
```
That looks like the right amount.

flar commented on engine/src/flutter/impeller/tessellator/tessellator.cc:
```diff
@@ -576,9 +584,17 @@ ArcVertexGenerator Tessellator::StrokedArc(const Matrix& view_transform,
+  if (cap == Cap::kRound) {
+    size_t round_cap_divisions = ComputeQuadrantDivisions(
+        view_transform.GetMaxBasisLengthXY() * half_width);
+    round_cap_trigs =
```
There is also a GetTrigsForDeviceRadius that does the `ComputeQuadrantDivisions` for you.

flar reviewed (commented):
This all looks correct, these are mostly just suggestions and I think you had some ideas about refactoring yourself, so I'm going to leave this as a "comment" review for now.

b-luk committed:
Updates

- Split up GenerateRoundCap into two functions for start/end
- Add factory constructors for ArcVertexGenerator
- Use GetTrigsForDeviceRadius instead of ComputeQuadrantDivisions + GetTrigsForDivisions

b-luk commented on engine/src/flutter/impeller/tessellator/tessellator.cc:
```diff
@@ -845,21 +890,32 @@ void Tessellator::GenerateRoundCapLine(
+      proc(p - relative_along + relative_across);
+      proc(p - relative_along - relative_across);
+    }
+  } else {
```
Done. Preserved the comment, and separated it into two calls. I added a bunch more documentation to the methods so hopefully that can help with clarity.

flar commented on engine/src/flutter/impeller/display_list/canvas.cc:
```diff
@@ -750,7 +750,7 @@ void Canvas::DrawArc(const Arc& arc, const Paint& paint) {
     // TODO(flar): It also only supports Butt and Square caps for now.
     // See https://github.com/flutter/flutter/issues/169400
-    if (oval_bounds.IsSquare() && paint.stroke.cap != Cap::kRound) {
+    if (oval_bounds.IsSquare()) {
```
You can delete the last 2 lines of the preceding comment (the todo line with the link to the issue) since you are fixing that bug.

flar commented on engine/src/flutter/impeller/tessellator/tessellator.h:
```diff
@@ -431,6 +448,43 @@ class Tessellator {
+  // directly opposite the semicircle's straight diameter. Vertices progress
+  // in the stroke's direction of travel, ending with two vertices at
+  // (p + perpendicular) and (p - perpendicular) that form the semicircle's
+  // diameter.
```
👍

flar reviewed (approved):
Looks good except to delete the obsolete part of that one comment.

flar commented:
The ninja problems are fixed at the top of tree. Update your branch with a pull (not a rebase) to fix those errors.

flar reviewed (commented):
I need to wait to see the impact on goldens before approving.

flar reviewed (changes requested):
It looks like the only way to remove an approval is to "Request Changes"... :(

b-luk committed:
Merge branch 'master' into arcstrokeroundcaps

b-luk committed:
Remove TODO comment

flar commented:
I'm guessing we'd see more improvements if the benchmark only measured the round cap case. Did you try hard-coding round caps in your benchmark code and testing before/after to see the isolated improvement?

b-luk commented:
Here are results after hard-coding the benchmark test to use round caps for every arc.

#### iPhone 12
|                       |          | before | after  | change  |
|-----------------------|----------|--------|--------|---------|
| frame build time      | avg      | 1.124  | 1.183  | 5.26%   |
|                       | 90th pct | 1.710  | 1.732  | 1.29%   |
|                       | 99th pct | 1.879  | 1.899  | 1.06%   |
|                       | worst    | 3.145  | 2.001  | -36.38% |
| rasterizer build time | avg      | 5.517  | 4.745  | -13.99% |
|                       | 90th pct | 11.640 | 6.303  | -45.85% |
|                       | 99th pct | 16.889 | 15.317 | -9.31%  |
|                       | worst    | 20.045 | 18.401 | -8.20%  |
| gpu frame time        | avg      | 10.800 | 10.509 | -2.69%  |
|                       | 90th pct | 15.625 | 15.625 | 0.00%   |
|                       | 99th pct | 15.625 | 15.625 | 0.00%   |
|                       | worst    | 23.438 | 15.625 | -33.33% |

#### Pixel 8
|                       |          | before | after  | change  |
|-----------------------|----------|--------|--------|---------|
| frame build time      | avg      | 6.403  | 6.094  | -4.82%  |
|                       | 90th pct | 9.103  | 8.307  | -8.74%  |
|                       | 99th pct | 10.995 | 10.187 | -7.35%  |
|                       | worst    | 11.335 | 11.436 | 0.89%   |
| rasterizer build time | avg      | 9.712  | 8.766  | -9.74%  |
|                       | 90th pct | 11.665 | 10.662 | -8.60%  |
|                       | 99th pct | 13.140 | 12.009 | -8.61%  |
|                       | worst    | 88.234 | 13.416 | -84.79% |

b-luk committed:
Merge branch 'master' into arcstrokeroundcaps

flar reviewed (approved):
LGTM

b-luk committed:
Merge branch 'master' into arcstrokeroundcaps
